src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java

```diff
@@ -81,28 +82,47 @@ void setFailureCount(int failureCount) {
     this.failureCount = failureCount;
   }
 
+  void setInvalidNodeStartTime(long invalidNodeStartTimeMillis) {
+    this.invalidNodeStartTime = invalidNodeStartTimeMillis;
+  }
+
+  void resetInvalidNodeStartTime() { this.invalidNodeStartTime = 0; }
+
+  boolean isInvalidNodeStartTimeDefined() { return this.invalidNodeStartTime > 0; }
+
+  public long getInvalidNodeStartTime() { return this.invalidNodeStartTime; }
+
   public boolean isNodeUnhealthy() {
     return this.isNodeUnhealthy;
   }
 
-  void updateConnectionStatus(long currentTime, boolean isValid, long connectionValidationElapsedTime) {
+  void updateConnectionStatus(long currentTime, boolean isValid, long validationIntervalTimeMillis) {
     final long totalElapsedTimeMillis = currentTime - this.startMonitorTime;
 
     if (totalElapsedTimeMillis > this.failureDetectionTimeMillis) {
-      this.setConnectionValid(isValid && (this.failureDetectionIntervalMillis >= connectionValidationElapsedTime));
+      this.setConnectionValid(isValid, currentTime, validationIntervalTimeMillis);
     }
   }
 
-  void setConnectionValid(boolean connectionValid) {
+  void setConnectionValid(boolean connectionValid, long currentTime, long validationIntervalTimeMillis) {
     if (!connectionValid) {
       this.failureCount++;
 
-      if (this.getFailureCount() >= this.getFailureDetectionCount()) {
+      if(!this.isInvalidNodeStartTimeDefined()) {
+        this.setInvalidNodeStartTime(currentTime);
+      }
+
+      long invalidNodeDurationMillis = currentTime - this.getInvalidNodeStartTime();
+      long maxInvalidNodeDurationMillis = this.getFailureDetectionIntervalMillis() * this.getFailureDetectionCount();
+      float adjustedFailureCount = this.getFailureDetectionIntervalMillis() / validationIntervalTimeMillis * this.getFailureDetectionCount();
+
+      // TODO: condition with failure counts may be unnecessary
+      if (this.getFailureCount() >= adjustedFailureCount && invalidNodeDurationMillis >= maxInvalidNodeDurationMillis) {
         this.log.logTrace(
```

Comment:
NIT: very complicated logic for logging one line. Not the goal of this PR, but a helper function would be good, so log lines are all one liners.
logger.info(message)